tests/test_agent.py

```diff
@@ -217,3 +224,286 @@ async def test_update_agent_map(server, environment, agent_factory):
+    python_code = f"""
+import json
+from inmanta.agent.agent import EnvBlueprint, PipConfig
+
+config_str = '''{config_str}'''
+config = json.loads(config_str)
+
+pip_config = PipConfig(**config["pip_config"])
+blueprint = EnvBlueprint(pip_config=pip_config, requirements=config["requirements"])
+
+# Generate and print the hash
+print(blueprint.generate_env_blueprint_hash())
+"""
+
+    # Generate hash in the current session for comparison
+    pip_config = PipConfig(**pip_config_dict)
+    current_session_blueprint = EnvBlueprint(pip_config=pip_config, requirements=requirements)
+    current_hash = current_session_blueprint.generate_env_blueprint_hash()
+
+    # Generate hash in a new interpreter session
+    result = subprocess.run(["python", "-c", python_code], capture_output=True, text=True)
+    new_session_hash = result.stdout.strip()
+
+    assert current_hash == new_session_hash, "Hash values should be consistent across interpreter sessions"
+
+
+async def test_environment_creation_locking(environment, tmpdir) -> None:
+    """
+    Tests the locking mechanism within VirtualEnvironmentManager to ensure that
+    only one environment is created for the same blueprint when requested concurrently,
+    preventing race conditions and duplicate environment creation.
+    """
+    manager = VirtualEnvironmentManager()
+
+    pip_index = PipIndex(artifact_dir=str(tmpdir))
+    create_python_package(
+        name="pkg1",
+        pkg_version=version.Version("1.0.0"),
+        path=os.path.join(tmpdir, "pkg1"),
+        publish_index=pip_index,
+    )
+
+    blueprint1 = EnvBlueprint(pip_config=PipConfig(index_url=pip_index.url), requirements=("pkg1",))
+    blueprint2 = EnvBlueprint(pip_config=PipConfig(index_url=pip_index.url), requirements=())
+
+    # Event to control the execution of get environment
+    creation_event = asyncio.Event()
+
+    # adjust the environment creation method to await the event
+    original_get_environment = manager.get_environment
+
+    async def mock_get_environment(blueprint, threadpool):
+        await creation_event.wait()  # Wait for the event to be set
+        return await original_get_environment(blueprint, threadpool)
+
+    manager.get_environment = mock_get_environment
+
+    # Start all get_environment but wait before proceeding
+    task_same_1 = asyncio.create_task(manager.get_environment(blueprint1, None))
+    task_same_2 = asyncio.create_task(manager.get_environment(blueprint1, None))
+    task_diff_1 = asyncio.create_task(manager.get_environment(blueprint2, None))
+
+    # Allow all get_environment tasks to proceed
+    creation_event.set()
+
+    # Wait for all tasks to complete
+    env_same_1, env_same_2, env_diff_1 = await asyncio.gather(task_same_1, task_same_2, task_diff_1)
+
+    assert env_same_1 is env_same_2, "Expected the same instance for the same blueprint"
+    assert env_same_1 is not env_diff_1, "Expected different instances for different blueprints"
```

Comment:
To which extend is this code different from the following code?
```suggestion

    # Wait for all tasks to complete
    env_same_1, env_same_2, env_diff_1 = await asyncio.gather(
        manager.get_environment(blueprint1, None),
        manager.get_environment(blueprint1, None),
        manager.get_environment(blueprint2, None),
    )

    assert env_same_1 is env_same_2, "Expected the same instance for the same blueprint"
    assert env_same_1 is not env_diff_1, "Expected different instances for different blueprints"
```

Reply:
I find my approach makes it more clear that the 3 calls are started at the same time.